pxtsim/simdriver.ts

```diff
@@ -120,16 +131,45 @@ namespace pxsim {
 
             this._allowedOrigins.push(this.getSimUrl().origin);
 
-            const messageSimulators = options?.messageSimulators
-            if (messageSimulators) {
-                Object.keys(messageSimulators)
-                    .map(channel => messageSimulators[channel])
-                    .forEach(messageSimulator => {
-                        this._allowedOrigins.push(new URL(messageSimulator.url).origin);
-                        if (messageSimulator.localHostUrl)
-                            this._allowedOrigins.push(new URL(messageSimulator.localHostUrl).origin);
-                    });
-            }
+            // Legacy support for message simulators
+            const messageSimulators = options?.messageSimulators || {};
+            Object.keys(messageSimulators)
+                .map(channel => messageSimulators[channel])
+                .forEach(messageSimulator => {
+                    this._allowedOrigins.push(new URL(messageSimulator.url).origin);
+                    if (messageSimulator.localHostUrl)
+                        this._allowedOrigins.push(new URL(messageSimulator.localHostUrl).origin);
+                });
+
+            // Preprocess simulator extensions
+            const simXDevMode = U.isLocalHost() && /[?&]simxdev(?:[=&#]|$)/i.test(window.location.href);
+            Object.entries(options?.simulatorExtensions || {}).forEach(([key, simx]) => {
+                // Verify essential `simx` config was provided
+                if (
+                    !simx ||
+                    !simx.index ||
+                    !simx.aspectRatio ||
+                    simx.permanent === undefined
+                ) {
+                    return;
+                }
+                // Compute the effective URL
+                if (simXDevMode && simx.devUrl) {
+                    // Use the dev URL if the dev flag is set (and we're on localhost)
+                    simx.url = new URL(simx.index, simx.devUrl).toString();
```

Comment:
This is interesting! It looks like `URL` gives some nice safeguards. I wonder if we should use this in more places.